platform-sdk/swirlds-jasperdb/src/main/java/com/swirlds/merkledb/collections/AbstractLongList.java

```diff
@@ -506,28 +509,72 @@ private void shrinkIfNeeded(final long newMinValidIndex) {
+        C chunk = chunkList.get(firstChunkWithDataIndex);
+        if (chunk != null && numberOfElementsToCleanUp > 0) {
+            partialChunkCleanup(chunk, true, numberOfElementsToCleanUp);
+        }
+
+        // clean up chunk(s) reserved for buffer
+        for (int i = firstValidChunkWithBuffer; i < firstChunkWithDataIndex; i++) {
+            chunk = chunkList.get(i);
+            if (chunk != null) {
+                partialChunkCleanup(chunk, true, numLongsPerChunk);
+            }
+        }
+    }
+
+    /**
+     * Deletes values from {@code maxValidIndex} to the end of the list and releases memory chunks reserved for these values.
+     * @param maxValidIndex new maximal valid index, right boundary of the list
+     */
+    private void shrinkRightSideIfNeeded(long maxValidIndex) {
+        final int listLength = chunkList.length();
+        final int lastValidChunkWithBufferIndex =
+                (int) Math.min((maxValidIndex + reservedBufferLength) / numLongsPerChunk, listLength - 1);
+        final int firstChunkIndexToDelete = lastValidChunkWithBufferIndex + 1;
+        final int numberOfChunks = calculateNumberOfChunks(size());
+
+        for (int i = firstChunkIndexToDelete; i < numberOfChunks; i++) {
+            final C chunk = chunkList.get(i);
+            if (chunk != null && chunkList.compareAndSet(i, chunk, null)) {
+                releaseChunk(chunk);
+            }
+        }
+
+        // clean up a chunk with data
+        final int firstChunkWithDataIndex = (int) (maxValidIndex / numLongsPerChunk);
+        final long numberOfElementsToCleanUp = numLongsPerChunk - (maxValidIndex % numLongsPerChunk) - 1;
+        C chunk = chunkList.get(firstChunkWithDataIndex);
+        if (chunk != null && numberOfElementsToCleanUp > 0) {
+            partialChunkCleanup(chunk, false, numberOfElementsToCleanUp);
+        }
 
         // clean up chunk(s) reserved for buffer
-        for (int i = firstValidChunkWithBuffer; i < firstChunkWithData; i++) {
-            partialChunkCleanup(i, numLongsPerChunk);
+        for (int i = firstChunkWithDataIndex + 1; i <= lastValidChunkWithBufferIndex; i++) {
+            chunk = chunkList.get(i);
+            if (chunk != null) {
+                partialChunkCleanup(chunk, false, numLongsPerChunk);
+            }
         }
     }
 
     /**
      * Releases a chunk that is no longer in use. Some implementation may preserve the chunk for further use.
      * @param chunk chunk to release
      */
-    protected abstract void releaseChunk(C chunk);
+    protected abstract void releaseChunk(@NonNull final C chunk);
 
     /**
-     * Zeroes out a part of a chunk
-     * @param chunkIndex index of the chunk to clean up
+     * Zeroes out a part of a chunk.
+     * @param chunk index of the chunk to clean up
+     * @param leftSide         if true, cleans up {@code elementsToCleanUp} on the left side of the chunk,
```

Comment:
`elementsToCleanUp` => `entriesToCleanUp`

The "...counting from the left boundary" comment below is stale.

Reply:
Fixed